cwms-data-api/src/main/java/cwms/cda/data/dao/LocationLevelsDaoImpl.java

```diff
@@ -540,5 +555,85 @@ private void getEnAttributeValue(StringValueUnits stringValueUnits, Double oattr
+        }
+        ZoneId locationZoneId = getLocationZoneId(levelRef.getLocationRef());
+        ZTSV_ARRAY specifiedTimes = buildTsvArray(start, end, interval, locationZoneId);
+        ZTSV_ARRAY locLvlValues = CWMS_LEVEL_PACKAGE.call_RETRIEVE_LOC_LVL_VALUES3(dsl.configuration(),
+                specifiedTimes, locationLevelId, levelUnits, attributeId, attributeValue, attributeUnits,
+                "UTC", officeId);
+        if (locLvlValues.isEmpty()) {
+            throw new NotFoundException("No time series found for: " + levelRef + " between start time: " + start + " and end time: " + end);
+        }
+        return buildTimeSeries(levelRef, interval, officeId, levelUnits, locLvlValues, locationZoneId);
+    }
 
+    private ZoneId getLocationZoneId(LocationTemplate locationRef) {
+        String timeZone = CWMS_LOC_PACKAGE.call_GET_LOCAL_TIMEZONE__2(dsl.configuration(),
+                locationRef.getLocationId(), locationRef.getOfficeId());
+        return OracleTypeMap.toZoneId(timeZone, locationRef.getLocationId());
+    }
+
+    private static TimeSeries buildTimeSeries(ILocationLevelRef levelRef, Interval interval, String officeId,
+                                              String levelUnits, ZTSV_ARRAY locLvlValues, ZoneId locationTimeZone) {
+        String timeSeriesId = levelRef.getLocationRef().getLocationId() + "." + levelRef.getParameter().getParameter()
+                + "." + levelRef.getParameterType().getParameterType() + "." + interval.getInterval() + "."
+                + levelRef.getDuration().toString() + "." + levelRef.getSpecifiedLevel().getId();
+        int size = locLvlValues.size();
+        Instant start = locLvlValues.get(0).getDATE_TIME().toInstant();
+        Instant end = locLvlValues.get(size - 1).getDATE_TIME().toInstant();
+        ZonedDateTime firstValueTime = ZonedDateTime.ofInstant(start, NumericalConstants.UTC_ZONEID);
+        ZonedDateTime lastValueTime = ZonedDateTime.ofInstant(end, NumericalConstants.UTC_ZONEID);
+        TimeSeries timeSeries = new TimeSeries(null, size, size, timeSeriesId,
+                officeId, firstValueTime, lastValueTime, levelUnits,
+                java.time.Duration.ofSeconds(interval.getSeconds()),
+                null, null, locationTimeZone.getId());
+        for (ZTSV_TYPE tsv : locLvlValues) {
+            Timestamp dateTime = tsv.getDATE_TIME();
+            Double value = tsv.getVALUE();
+            if (value == null) {
+                value = NumericalConstants.HEC_UNDEFINED_DOUBLE;
+            }
+            BigDecimal qualityCode = tsv.getQUALITY_CODE();
+            int quality = 0;
+            if (qualityCode != null) {
+                quality = qualityCode.intValue();
+            }
+            timeSeries.addValue(dateTime, value, quality);
+        }
+        return timeSeries;
+    }
+
+    private ZTSV_ARRAY buildTsvArray(Instant start, Instant end, Interval interval, ZoneId locationTimeZone) {
```

Comment:
maybe just this part feel duplicated. The above is fairly specific to getting from the location level.

Maybe it's also just the names that are the same.

Reply:
The difference here is that the time series dao leverages the AV_TSV view generating the time steps for us. For the levels we generate the time step in CDA and request the values/quality from CWMS.

I could see an iteration on this design where we passed in an array of times in the GET body and skip interval processing at the CDA side altogether. I opted for CDA doing the processing as it would be less of a payload sent from the client and I didn't want to get into the complication of whether we should even be adding a body to a GET request.

Reply:
Oh, I see. that make sense. We can revisit later to see if there is improved design.

Reply:
updated